src/inmanta/data/__init__.py

```diff
@@ -4203,6 +4277,116 @@ async def query_resource_actions(
+        3. get all resources required by this resource
+        4. get all resource actions of type deploy emitted by the resource of step 3 in the time interval of step 2
+
+        :param env: environment to consider
+        :param resource_id: resource to consider, should be in deploying state
+        :param exclude_change: in step 4, exclude all resource actions with this specific type of change
+        """
+
+        # This is bang on the critical path for the agent
+        # Squeeze out as much performance from postgresql as we can
+
+        # steps 1 and 2:
+        # find the interval between the current deploy and the previous successful deploy
+        # also check we are currently deploying
+        # do all of this in one query
+        resource_id_str = resource_id.resource_version_str()
+
+        # These two variables are actually of type datetime.datetime
+        # but mypy doesn't know as they come from the DB
+        # mypy also doesn't care, because they go back into the DB
+        current_deploy_start: object
+        last_deploy_start: Optional[object]
+
+        end_query = """
+        with
+            base_ra as (
+            SELECT ra.*
+                FROM public.resourceaction_resource as jt
+                    INNER JOIN public.resourceaction as ra
+                        ON ra.action_id = jt.resource_action_id
+                    WHERE jt.environment=$1 AND ra.environment=$1 AND jt.resource_id=$2::varchar AND ra.action='deploy'
+                    ORDER BY ra.started DESC
+            )
+        SELECT
+            (SELECT started from base_ra ORDER BY started DESC LIMIT 1) as begin_started,
+            (SELECT status from base_ra ORDER BY started DESC LIMIT 1) as begin_status,
+            COALESCE((SELECT started from base_ra where status='deployed' ORDER BY started DESC LIMIT 1), NULL) as started
+        """
+
+        async with cls.get_connection() as connection:
+            result = await connection.fetchrow(end_query, env.id, resource_id.resource_str())
+
+            if not result or result["begin_status"] is None:
+                raise BadRequest(
+                    "Fetching resource events only makes sense when the resource is currently deploying. Resource"
+                    f" {resource_id_str} has not started deploying yet."
+                )
+            if result["begin_status"] != const.ResourceState.deploying:
+                raise BadRequest(
+                    "Fetching resource events only makes sense when the resource is currently deploying. Current deploy state"
+                    f" for resource {resource_id_str} is {result['begin_status']}."
+                )
+            current_deploy_start = result["begin_started"]
+            last_deploy_start = result["started"]
+
+            # Step3: Get the resource
+            resource: Optional[Resource] = await Resource.get_one(environment=env.id, resource_version_id=resource_id_str)
```

Comment:
This is not run on the same connection, is that intentional? Since you don't start a transaction, any reason you create a connection in the first place rather than just using `cls._fetchrow()` and `cls._fetchquery()`?

Reply:
That's an interesting observation. I was wondering about this yesterday but I don't think I saw the full implications in this context. You could maybe even deadlock (not PG deadlock but Python deadlock) on this: N methods calls (to the same or different methods) having acquired a lock, then waiting to acquire another one for a query for which the connection object wasn't wired through but that lives (as far as Python is concerned) within the context of the acquired connection.